### PR TITLE
csi snapshot

### PR DESCRIPTION
**Pull request by BaiHuoYu**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
Csi snapshot
**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: csi/server/deploy/kubernetes/csi-snapshotter-rbac.yaml [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# This YAML file contains RBAC API objects,
+# which are necessary to run external csi provisioner for opensds.
```
Comment (xing-yang) [inline csi/server/deploy/kubernetes/csi-snapshotter-rbac.yaml]: s/provisioner/snapshotter

File: csi/server/plugin/opensds/controller.go [hunk 2]
```diff
@@ -451,30 +453,125 @@ func (p *Plugin) ControllerGetCapabilities(
...
-// CreateSnapshot
+// CreateSnapshot implementation
 func (p *Plugin) CreateSnapshot(
 	ctx context.Context,
 	req *csi.CreateSnapshotRequest) (
 	*csi.CreateSnapshotResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+
+	defer glog.Info("end to CreateSnapshot")
+	glog.Infof("start to CreateSnapshot, Name: %v, SourceVolumeId: %v, CreateSnapshotSecrets: %v, parameters: %v!",
```
Comment (xing-yang) [inline csi/server/plugin/opensds/controller.go]: Detailed debug log msg should be glog.V(5).Infof

File: csi/server/plugin/opensds/controller.go [hunk 3]
```diff
@@ -451,30 +453,125 @@ func (p *Plugin) ControllerGetCapabilities(
...
-// DeleteSnapshot
+// DeleteSnapshot implementation
 func (p *Plugin) DeleteSnapshot(
 	ctx context.Context,
 	req *csi.DeleteSnapshotRequest) (
 	*csi.DeleteSnapshotResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+
+	defer glog.Info("end to DeleteSnapshot")
+	glog.Infof("start to DeleteSnapshot, SnapshotId: %v, DeleteSnapshotSecrets: %v!",
```
Comment (xing-yang) [inline csi/server/plugin/opensds/controller.go]: V(5)

File: csi/server/plugin/opensds/controller.go [hunk 4]
```diff
@@ -451,30 +453,125 @@ func (p *Plugin) ControllerGetCapabilities(
 					},
 				},
 			},
+			&csi.ControllerServiceCapability{
+				Type: &csi.ControllerServiceCapability_Rpc{
+					Rpc: &csi.ControllerServiceCapability_RPC{
+						Type: csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT,
+					},
+				},
+			},
```
Comment (xing-yang) [inline csi/server/plugin/opensds/controller.go]: Need to add capablity for LIST_SNAPSHOTS too

File: csi/server/plugin/opensds/controller.go [hunk 5]
```diff
@@ -451,30 +453,125 @@ func (p *Plugin) ControllerGetCapabilities(
...
-// ListSnapshots
+// ListSnapshots implementation
 func (p *Plugin) ListSnapshots(
 	ctx context.Context,
 	req *csi.ListSnapshotsRequest) (
 	*csi.ListSnapshotsResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+
+	defer glog.Info("end to ListSnapshots")
+	glog.Infof("start to ListSnapshots, MaxEntries: %v, StartingToken: %v, SourceVolumeId: %v, SnapshotId: %v!",
```
Comment (xing-yang) [inline csi/server/plugin/opensds/controller.go]: V(5)

File: csi/server/plugin/opensds/controller_test.go [hunk 6]
```diff
@@ -87,6 +87,13 @@ func TestControllerGetCapabilities(t *testing.T) {
 				},
 			},
 		},
+		&csi.ControllerServiceCapability{
+			Type: &csi.ControllerServiceCapability_Rpc{
+				Rpc: &csi.ControllerServiceCapability_RPC{
+					Type: csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT,
+				},
+			},
+		},
```
Comment (xing-yang) [inline csi/server/plugin/opensds/controller_test.go]: Need to add more unit tests

File: csi/server/plugin/opensds/controller.go [hunk 7]
```diff
@@ -451,30 +453,125 @@ func (p *Plugin) ControllerGetCapabilities(
...
+	}
+
+	createdAt, err := time.Parse(constants.TimeFormat, snapshot.CreatedAt)
+	if nil != err {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &csi.CreateSnapshotResponse{
+		Snapshot: &csi.Snapshot{
+			SizeBytes:      snapshot.Size,
```
Comment (xing-yang) [inline csi/server/plugin/opensds/controller.go]: Size seems to be wrong from your video.  You created a "1Gi" volume specified in pvc_sc.yaml, but "kubectl describe volumesnapshot" showed "Restore Size: 1".  That means 1 byte?   It should be "Restore Size:  1Gi".

**Comment by xing-yang** [comment]
Please add unit tests.

File: csi/server/plugin/opensds/controller.go (hunk 4 above)
Comment (BaiHuoYu) [reply csi/server/plugin/opensds/controller.go]: LIST_SNAPSHOTS has been added

File: csi/server/plugin/opensds/controller_test.go (hunk 6 above)
Comment (BaiHuoYu) [reply csi/server/plugin/opensds/controller_test.go]: Unit tests have been added

**Pull request merged**